sdks/python/apache_beam/io/gcp/spannerio.py

```diff
@@ -0,0 +1,531 @@
+    self._snapshot = self._database.batch_snapshot(**self._snapshot_options)
+    return [self._snapshot.to_dict()]
+
+  def teardown(self):
+    if self._snapshot:
+      self._snapshot.close()
+
+
+@ptransform_fn
+def create_transaction(pbegin, project_id, instance_id, database_id,
+                       credentials=None, pool=None, read_timestamp=None,
+                       exact_staleness=None):
+  """
+  A PTransform method to create a batch transaction.
+
+  Args:
+    pbegin: Root of the pipeline
+    project_id: Cloud spanner project id. Be sure to use the Project ID,
+      not the Project Number.
+    instance_id: Cloud spanner instance id.
+    database_id: Cloud spanner database id.
+    credentials: (optional) The authorization credentials to attach to requests.
+      These credentials identify this application to the service.
+      If none are specified, the client will attempt to ascertain
+      the credentials from the environment.
+    pool: (optional) session pool to be used by database. If not passed, the
+      database will construct an instance of BurstyPool.
+    read_timestamp: (optional) An instance of the `datetime.datetime` object to
+      execute all reads at the given timestamp.
+    exact_staleness: (optional) And instance of `datetime.timedelta` object to
+      execute all reads at a timestamp that is exact_staleness old.
+  """
+
+  assert isinstance(pbegin, PBegin)
+
+  return (pbegin | Create([1]) | ParDo(_CreateTransactionFn(
+      project_id, instance_id, database_id, credentials,
+      pool, read_timestamp,
+      exact_staleness)))
+
+
+class _ReadFromPartitionFn(DoFn):
+  """
+  A DoFn to perform reads from the partition.
+  """
+
+  def __init__(self, spanner_configuration):
+    self._spanner_configuration = spanner_configuration
+
+  def to_runner_api_parameter(self, context):
+    return self.to_runner_api_pickled(context)
+
+  def setup(self):
+    spanner_client = Client(self._spanner_configuration.project)
+    instance = spanner_client.instance(self._spanner_configuration.instance)
+    self._database = instance.database(self._spanner_configuration.database,
+                                       pool=self._spanner_configuration.pool)
+    self._snapshot = self._database.batch_snapshot(**self._spanner_configuration
+                                                   .snapshot_options)
+
+  def process(self, element):
+    self._snapshot = BatchSnapshot.from_dict(
+        self._database,
+        element['transaction_info']
+    )
+
+    if element['read_operation'] == 'process_query_batch':
+      read_action = self._snapshot.process_query_batch
+    elif element['read_operation'] == 'process_read_batch':
+      read_action = self._snapshot.process_read_batch
+    else:
+      raise ValueError("Unknown read action.")
+
+    for row in read_action(element['partitions']):
+      yield row
+
+  def teardown(self):
+    if self._snapshot:
+      self._snapshot.close()
```

Comment:
This needs an integration test before we can be sure it works as expected.
For example, I'm not sure if closing the BatchSnapshot here closes the read transaction.

Reply:
The jira mentions that the IT and performance testing will be on a separate ticket!
https://issues.apache.org/jira/browse/BEAM-7246

Reply:
Pleas change the classes here to private (for example, _ReadFromSpanner) till we have proper integration tests in place to prevent users from using a solution without proper testing in production. Also create a separate ticket for integration tests and refer to it here.

Having some integration tests in place is a must before making this publicly available IMHO. Performance tests are also helpful but optional.

Reply:
Hi @chamikaramj. Thanks for the feedback. Yea IT tests should be there to make sure it's working properly. It's a good thought to make this transform private while some tests is remaining. 

I have made the changes you mentioned. And also created the ticket for the IT tests.
[BEAM-8949](https://issues.apache.org/jira/browse/BEAM-8949)